src/com/sun/jna/win32/W32StringUtil.java

```diff
@@ -0,0 +1,390 @@
+        Memory pointer = new Memory(buffer.length + 1 * getCharWidth());
+        pointer.write(0, buffer, 0, buffer.length);
+
+        // Ensure null-termination
+        byte[] zeros = new byte[getCharWidth()];
+        Arrays.fill(zeros, (byte)0);
+        pointer.write(buffer.length, zeros, 0, zeros.length);
+
+        if (isAPITypeWide()) {
+            return pointer.getWideString(offsetInCharacters * getCharWidth());
+        } else {
+            return pointer.getString(offsetInCharacters);
+        }
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @return The string
+     * @see #toString(char[], int)
+     * @see #toString(char[], int, int)
+     */
+    public static String toString(char[] buffer) {
+        return toString(buffer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(char[])
+     * @see #toString(char[], int, int)
+     */
+    public static String toString(char[] buffer, int offsetInCharacters) {
+        Memory pointer = new Memory((buffer.length + 1) * Native.WCHAR_SIZE); // Always allocate for wide
+        pointer.write(0, buffer, 0, buffer.length);
+
+        // Ensure null-termination
+        byte[] zeros = new byte[getCharWidth()];
+        Arrays.fill(zeros, (byte)0);
+        pointer.write(buffer.length * getCharWidth(), zeros, 0, zeros.length);
+
+        if (isAPITypeWide()) {
+            return pointer.getWideString(offsetInCharacters * getCharWidth());
+        } else {
+            return pointer.getString(offsetInCharacters);
+        }
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @param lengthInCharacters The max length in characters of the string
+     * @return The string
+     * @see #toString(char[])
+     * @see #toString(char[], int)
+     */
+    public static String toString(char[] buffer, int offsetInCharacters, int lengthInCharacters) {
+        return toString(buffer, offsetInCharacters).substring(0, lengthInCharacters);
+    }
+
+    /**
+     * Convert buffer to a raw string
+     *
+     * @param biffer The buffer
+     * @return The string. Note that the string can contain embedded \0.
+     * @see #toRawString(char[])
+     */
+    public static String toRawString(char[] buffer) {
+        if (isAPITypeWide()) {
+            return new String(buffer);
+        } else {
+            Memory pointer = new Memory(buffer.length * Native.WCHAR_SIZE); // Always allocate for wide
+            pointer.write(0, buffer, 0, buffer.length);
+            byte[] bytes = pointer.getByteArray(0, (int)pointer.size());
+            return new String(bytes, 0, buffer.length); // Only the first part of the buffer contains data
```

Comment:
I don't understand this code. Creating a string from raw bytes without a charset is broken nearly everytime.

Reply:
True, I'll add the call to Native.getDefaultStringEncoding().